## **H1N1 AND SEASONAL FLU VACCINES** ##

**BUSINESS UNDERSTANDING**


The COVID-19 pandemic has highlighted the critical importance of vaccination campaigns in controlling the spread of infectious diseases. Public health efforts are more effective when they are informed by an understanding of the factors that influence individuals' decisions to get vaccinated. By analyzing past vaccination patterns, such as those during the 2009 H1N1 flu pandemic, we can gain valuable insights that can help design better vaccination strategies, improve communication campaigns, and ultimately increase vaccine uptake in future public health crises.

**PROBLEM STATEMENT**

The task is to predict whether individuals received the H1N1 flu vaccine or the seasonal flu vaccine using data from the 2009 National H1N1 Flu Survey. This binary classification problem involves analyzing various factors, such as demographics, opinions, and health behaviors, to determine their relationship with vaccination behavior. The outcome of this analysis will help public health authorities identify key factors influencing vaccine acceptance and inform strategies to enhance vaccine coverage in the population.

**OBJECTIVES**

**Identify Key Predictors**: Determine the most significant factors that influence whether an individual received the H1N1 or seasonal flu vaccine in 2009.

**Develop a Predictive Model**: Build a binary classification model to accurately predict whether a survey respondent received the chosen vaccine (either H1N1 or seasonal flu).

**Evaluate Model Performance**: Assess the model's performance using appropriate metrics such as accuracy, precision, recall, and F1 score to ensure its reliability in predicting vaccination behavior.

**Provide Actionable Insights**: Analyze the model's findings to provide public health authorities with actionable insights that can guide future vaccination campaigns and strategies, particularly in the context of managing public health responses to pandemics.

## **DATA UNDERSTANDING** ##

**Importing Libraries**

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

**Loading Datasets**

We will load the train, test and label datasets. The train will be used to train our model, the test will be used for  testing the model performance and  the label dataset which has the following target variables:
h1n1_vaccine - Whether respondent received H1N1 flu vaccine.
seasonal_vaccine - Whether respondent received seasonal flu vaccine.

In [4]:
train_data = pd.read_csv('Data/training_set_features.csv')
test_data = pd.read_csv('Data/test_set_features.csv')
label_data = pd.read_csv('Data/training_set_labels.csv')

**Data Inspection**

In [11]:
#inspecting the data sizes of the train,test and label
print('Train shape:', train_data.shape)

print('test shape:', test_data.shape)

print('label shape:', label_data.shape)

Train shape: (26707, 36)
test shape: (26708, 36)
label shape: (26707, 3)


From above results, we see that both the train and test datasets have the same number of rows(26,707) and columns(36). The label data on the other end has also same number of rows which is 26,707 and 3 columns. This is because the label data just contains the target variables and the responded ids.

In [8]:
#inspecting the first five rows of the train dataset
train_data.head()

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0              0           1.0             0.0                        0.0   
1              1           3.0             2.0                        0.0   
2              2           1.0             1.0                        0.0   
3              3           1.0             1.0                        0.0   
4              4           2.0             1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...             income_poverty  marital_status  \
0                    1.0  ...              Below Poverty     Not Married   
1                    1.0  ...              Below Poverty     Not Married   
2                    0.0  ...  <= $75,000, Above Poverty     Not Married   
3                    0.0  ...              Below Poverty     Not Married   
4                    1.0  ...  <= $75,000, Above Poverty         Married   

   rent_or_own   employment_status  hhs_geo_region                census_msa  \
0          Own  Not in Labor Force        oxchjgsf                   Non-MSA   
1         Rent            Employed        bhuqouqj  MSA, Not Principle  City   
2          Own            Employed        qufhixun  MSA, Not Principle  City   
3         Rent  Not in Labor Force        lrircsnp       MSA, Principle City   
4          Own            Employed        qufhixun  MSA, Not Principle  City   

   household_adults  household_children  employment_industry  \
0               0.0                 0.0                  NaN   
1               0.0                 0.0             pxcmvdjn   
2               2.0                 0.0             rucpziij   
3               0.0                 0.0                  NaN   
4               1.0                 0.0             wxleyezf   

   employment_occupation  
0                    NaN  
1               xgwztkwe  
2               xtkaffoo  
3                    NaN  
4               emcorrxb  

[5 rows x 36 columns]

In [9]:
#inspecting the first five rows of the test dataset
test_data.head()

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0          26707           2.0             2.0                        0.0   
1          26708           1.0             1.0                        0.0   
2          26709           2.0             2.0                        0.0   
3          26710           1.0             1.0                        0.0   
4          26711           3.0             1.0                        1.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   1.0                   0.0                    1.0   
1                   0.0                   0.0                    0.0   
2                   0.0                   1.0                    1.0   
3                   0.0                   0.0                    0.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          1.0                      0.0   
1                          0.0                      0.0   
2                          1.0                      1.0   
3                          0.0                      0.0   
4                          1.0                      1.0   

   behavioral_touch_face  ...             income_poverty  marital_status  \
0                    1.0  ...                  > $75,000     Not Married   
1                    0.0  ...              Below Poverty     Not Married   
2                    1.0  ...                  > $75,000         Married   
3                    0.0  ...  <= $75,000, Above Poverty         Married   
4                    1.0  ...  <= $75,000, Above Poverty     Not Married   

   rent_or_own   employment_status  hhs_geo_region                census_msa  \
0         Rent            Employed        mlyzmhmf  MSA, Not Principle  City   
1         Rent            Employed        bhuqouqj                   Non-MSA   
2          Own            Employed        lrircsnp                   Non-MSA   
3          Own  Not in Labor Force        lrircsnp  MSA, Not Principle  City   
4          Own            Employed        lzgpxyit                   Non-MSA   

   household_adults  household_children  employment_industry  \
0               1.0                 0.0             atmlpfrs   
1               3.0                 0.0             atmlpfrs   
2               1.0                 0.0             nduyfdeo   
3               1.0                 0.0                  NaN   
4               0.0                 1.0             fcxhlnwr   

   employment_occupation  
0               hfxkjkmi  
1               xqwwgdyp  
2               pvmttkik  
3                    NaN  
4               mxkfnird  

[5 rows x 36 columns]

From the first five rows for both the train and test, we can note that some columns have missing values as well as data that does not have any significant meaning for utilization. This is observed in the employment_industry, employment_occupation and the hse_geo_region features. We will further handle this later during preprocessing.

In [10]:
label_data.head()

respondent_id  h1n1_vaccine  seasonal_vaccine
0              0             0                 0
1              1             0                 1
2              2             0                 0
3              3             0                 1
4              4             0                 0

something woth noting from the inspection of the label data is that it conatains  a key column which is the respondent_id which was also available in the train dataset. We will use this column to marge the target variables to the train data to aid with our modeling process.

In [12]:
#overall inspection of the dataset
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26707 entries, 0 to 26706
Data columns (total 36 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   respondent_id                26707 non-null  int64  
 1   h1n1_concern                 26615 non-null  float64
 2   h1n1_knowledge               26591 non-null  float64
 3   behavioral_antiviral_meds    26636 non-null  float64
 4   behavioral_avoidance         26499 non-null  float64
 5   behavioral_face_mask         26688 non-null  float64
 6   behavioral_wash_hands        26665 non-null  float64
 7   behavioral_large_gatherings  26620 non-null  float64
 8   behavioral_outside_home      26625 non-null  float64
 9   behavioral_touch_face        26579 non-null  float64
 10  doctor_recc_h1n1             24547 non-null  float64
 11  doctor_recc_seasonal         24547 non-null  float64
 12  chronic_med_condition        25736 non-null  float64
 13  child_under_6_mo

The dataset contains 26,707 entries with 36 columns.The dataset is a mix of numerical and categorical data types, with 23 columns of type float64, 1 column of type int64, and 12 columns of type object.

**Key Observations**:

**Missing Data**:

Several columns contain missing values, which need to be addressed during the data cleaning process. For example:

doctor_recc_h1n1 and doctor_recc_seasonal have a significant amount of missing data, with 24,547 non-null values out of 26,707 entries.

health_insurance has missing values in over 45% of the entries, with only 14,433 non-null values.

The columns employment_industry and employment_occupation have the highest number of missing values, with only 13,377 and 13,237 non-null values, respectively.

**Column Types**:
Based on the data types observed for each column types described below, we will not need to do any data type conversion since sll the columns are in their desired datatype format.
**Numeric Columns:** There are 23 numeric columns (float64 type) that primarily represent levels of concern, knowledge, and behaviors related to H1N1 and seasonal flu, as well as some demographic information (e.g., household size).
**Categorical Columns:** The dataset includes 12 categorical columns (object type), such as age_group, education, race, sex, and employment_status. These are critical for demographic analysis and will likely need encoding before modeling.

**Identifier Column:** respondent_id is a unique identifier for each survey respondent and is of type int64.

In [13]:
#overall statistics
train_data.describe()

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
count   26707.000000  26615.000000    26591.000000               26636.000000   
mean    13353.000000      1.618486        1.262532                   0.048844   
std      7709.791156      0.910311        0.618149                   0.215545   
min         0.000000      0.000000        0.000000                   0.000000   
25%      6676.500000      1.000000        1.000000                   0.000000   
50%     13353.000000      2.000000        1.000000                   0.000000   
75%     20029.500000      2.000000        2.000000                   0.000000   
max     26706.000000      3.000000        2.000000                   1.000000   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
count          26499.000000          26688.000000           26665.000000   
mean               0.725612              0.068982               0.825614   
std                0.446214              0.253429               0.379448   
min                0.000000              0.000000               0.000000   
25%                0.000000              0.000000               1.000000   
50%                1.000000              0.000000               1.000000   
75%                1.000000              0.000000               1.000000   
max                1.000000              1.000000               1.000000   

       behavioral_large_gatherings  behavioral_outside_home  \
count                  26620.00000             26625.000000   
mean                       0.35864                 0.337315   
std                        0.47961                 0.472802   
min                        0.00000                 0.000000   
25%                        0.00000                 0.000000   
50%                        0.00000                 0.000000   
75%                        1.00000                 1.000000   
max                        1.00000                 1.000000   

       behavioral_touch_face  ...  health_worker  health_insurance  \
count           26579.000000  ...   25903.000000       14433.00000   
mean                0.677264  ...       0.111918           0.87972   
std                 0.467531  ...       0.315271           0.32530   
min                 0.000000  ...       0.000000           0.00000   
25%                 0.000000  ...       0.000000           1.00000   
50%                 1.000000  ...       0.000000           1.00000   
75%                 1.000000  ...       0.000000           1.00000   
max                 1.000000  ...       1.000000           1.00000   

       opinion_h1n1_vacc_effective  opinion_h1n1_risk  \
count                 26316.000000       26319.000000   
mean                      3.850623           2.342566   
std                       1.007436           1.285539   
min                       1.000000           1.000000   
25%                       3.000000           1.000000   
50%                       4.000000           2.000000   
75%                       5.000000           4.000000   
max                       5.000000           5.000000   

       opinion_h1n1_sick_from_vacc  opinion_seas_vacc_effective  \
count                 26312.000000                 26245.000000   
mean                      2.357670                     4.025986   
std                       1.362766                     1.086565   
min                       1.000000                     1.000000   
25%                       1.000000                     4.000000   
50%                       2.000000                     4.000000   
75%                       4.000000                     5.000000   
max                       5.000000                     5.000000   

       opinion_seas_risk  opinion_seas_sick_from_vacc  household_adults  \
count       26193.000000                 26170.000000      26458.000000   
mean            2.719162                     2.118112          0.886499   
std             1.385055                     1.332

**Binary Variables (0 or 1 values):**

Many columns, such as behavioral_antiviral_meds, behavioral_avoidance, behavioral_face_mask, and others, have minimum values of 0 and maximum values of 1, which are typical for binary variables. However, the mean values for these variables are quite low, suggesting that most respondents answered 0 (indicating 'No'). Since the range is limited to 0 and 1, no outliers can be identified in these columns based on the max, min, or mean values.

**Ordinal Variables (e.g., h1n1_concern, h1n1_knowledge):**

Variables like h1n1_concern and h1n1_knowledge have minimum values of 0, maximum values of 3 or 2, and mean values that do not reach the maximum. This distribution suggests a skew towards lower concern or knowledge, with fewer respondents scoring at the high end. Since these are ordinal variables, the values themselves are constrained, but if a variable has a mean much closer to the lower end (e.g., h1n1_knowledge with a mean of 1.26 and a max of 2), it suggests a concentration of responses at the lower levels.


**Household Variables (household_adults, household_children):**

Household variables, such as household_adults and household_children, typically have minimum values of 0 (indicating no adults or children in the household) and maximum values that reflect the upper limits of the survey data. For instance, a higher max value for household_adults could indicate larger family sizes in some households. However, the mean values for these variables tend to be closer to the lower end, suggesting that most households have fewer adults or children. In these cases, extreme values (such as a very high number of adults or children) could be considered outliers if they deviate significantly from the typical household structure, but these outliers might reflect genuine variability in household sizes rather than errors in the data.

In [14]:
#getting the column names
train_data.columns

Index(['respondent_id', 'h1n1_concern', 'h1n1_knowledge',
       'behavioral_antiviral_meds', 'behavioral_avoidance',
       'behavioral_face_mask', 'behavioral_wash_hands',
       'behavioral_large_gatherings', 'behavioral_outside_home',
       'behavioral_touch_face', 'doctor_recc_h1n1', 'doctor_recc_seasonal',
       'chronic_med_condition', 'child_under_6_months', 'health_worker',
       'health_insurance', 'opinion_h1n1_vacc_effective', 'opinion_h1n1_risk',
       'opinion_h1n1_sick_from_vacc', 'opinion_seas_vacc_effective',
       'opinion_seas_risk', 'opinion_seas_sick_from_vacc', 'age_group',
       'education', 'race', 'sex', 'income_poverty', 'marital_status',
       'rent_or_own', 'employment_status', 'hhs_geo_region', 'census_msa',
       'household_adults', 'household_children', 'employment_industry',
       'employment_occupation'],
      dtype='object')

In [15]:
#checking the unique values for some features to determine how to later handle them during preprocessing

unique_values = {
    'income_poverty': train_data['income_poverty'].unique(),
    'marital_status': train_data['marital_status'].unique(),
    'rent_or_own': train_data['rent_or_own'].unique(),
    'employment_status': train_data['employment_status'].unique(),
    'education': train_data['education'].unique(),
    'race': train_data['race'].unique(),
    'sex': train_data['sex'].unique(),
    'age_group': train_data['age_group'].unique(),
    'train_data.columns': train_data.columns.tolist()
}

# Print unique values for each feature
for feature, values in unique_values.items():
    print(f"Unique values for {feature}: {values}")

Unique values for income_poverty: ['Below Poverty' '<= $75,000, Above Poverty' '> $75,000' nan]
Unique values for marital_status: ['Not Married' 'Married' nan]
Unique values for rent_or_own: ['Own' 'Rent' nan]
Unique values for employment_status: ['Not in Labor Force' 'Employed' 'Unemployed' nan]
Unique values for education: ['< 12 Years' '12 Years' 'College Graduate' 'Some College' nan]
Unique values for race: ['White' 'Black' 'Other or Multiple' 'Hispanic']
Unique values for sex: ['Female' 'Male']
Unique values for age_group: ['55 - 64 Years' '35 - 44 Years' '18 - 34 Years' '65+ Years'
 '45 - 54 Years']
Unique values for train_data.columns: ['respondent_id', 'h1n1_concern', 'h1n1_knowledge', 'behavioral_antiviral_meds', 'behavioral_avoidance', 'behavioral_face_mask', 'behavioral_wash_hands', 'behavioral_large_gatherings', 'behavioral_outside_home', 'behavioral_touch_face', 'doctor_recc_h1n1', 'doctor_recc_seasonal', 'chronic_med_condition', 'child_under_6_months', 'health_worker', 'h

From above inspection, we observe the following:

Income Poverty: Categories include 'Below Poverty', '<= $75,000, Above Poverty', '> $75,000', with some missing values (NaN).
Marital Status: Includes 'Not Married', 'Married', with missing values.
Rent or Own: 'Own' vs. 'Rent', with missing values.
Employment Status: 'Not in Labor Force', 'Employed', 'Unemployed', and some missing values.
Education: 'Below 12 Years', '12 Years', 'Some College', 'College Graduate', with missing values.
Race: 'White', 'Black', 'Other or Multiple', 'Hispanic'.
Sex: 'Female', 'Male'.
Age Group: '18 - 34 Years', '35 - 44 Years', '45 - 54 Years', '55 - 64 Years', '65+ Years'.

From this we can decide to impute our missing values with 'unknown' value to avoid any manipulation or we can decide to drop them if the impact is minimal